data-prepper-plugins/otel-trace-raw-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltrace/OTelTraceRawPrepper.java

```diff
@@ -94,11 +120,122 @@ public OTelTraceRawPrepper(final PluginSetting pluginSetting) {
                 }
             }
         }
-        return finalRecords;
+
+        final List<RawSpan> rawSpans = new LinkedList<>();
+        rawSpans.addAll(getTracesToFlushByParentSpan());
+        rawSpans.addAll(getTracesToFlushByGarbageCollection());
+
+        return convertRawSpansToJsonRecords(rawSpans);
+    }
+
+    private List<Record<String>> convertRawSpansToJsonRecords(final List<RawSpan> rawSpans) {
+        final List<Record<String>> records = new LinkedList<>();
+
+        for (RawSpan rawSpan : rawSpans) {
+            String rawSpanJson;
+            try {
+                rawSpanJson = rawSpan.toJson();
+            } catch (JsonProcessingException e) {
+                log.error("Unable to process invalid Span {}:", rawSpan, e);
+                spanErrorsCounter.increment();
+                totalProcessingErrorsCounter.increment();
+                continue;
+            }
+
+            records.add(new Record<>(rawSpanJson));
+        }
+
+        return records;
+    }
+
+    private List<RawSpan> getTracesToFlushByParentSpan() {
+        final List<RawSpan> recordsToFlush = new LinkedList<>();
+
+        for (DelayedParentSpan delayedParentSpan; (delayedParentSpan = delayedParentSpanQueue.poll()) != null;) {
+            RawSpan parentSpan = delayedParentSpan.getRawSpan();
+            recordsToFlush.add(parentSpan);
+
+            String traceGroup = parentSpan.getTraceGroup();
+            String parentSpanTraceId = parentSpan.getTraceId();
+
+            RawSpanSet rawSpanSet = traceIdRawSpanSetMap.get(parentSpanTraceId);
```

Comment:
I'm considering instead of getting the parent span from the queue, and the child spans from the set, if we were to write the parent span into the set with the children on ingest and then just grab the whole trace from the span set when it's time to flush. 

Might be simpler to understand, and also if the Map of Sets is backed by MapDB on disk then it would make sense keeping the parent with the children. In this approach we'd just get the traceId from the parent span, then use it to fetch _everything_ from the map of sets.

Reply:
Discussed offline. We reserve off-heap implementation for the next iteration depending on heap memory consumption in the load test. Also resilience to instance outage will be taken into account.

Reply:
Sure, can revisit later.